controller/k8s/metadata_api.go

```diff
@@ -0,0 +1,256 @@
+	ls, err := api.getLister(res)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := ls.ByNamespace(ns).Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	nsMeta, ok := obj.(*metav1.PartialObjectMetadata)
+	if !ok {
+		return nil, fmt.Errorf("couldn't convert obj %v to PartialObjectMetadata", obj)
+	}
+
+	return nsMeta, nil
+}
+
+// GetByNamespaceFilteredCached returns a list of Kubernetes object references,
+// given a type, namespace, name and label selector
+func (api *MetadataAPI) GetByNamespaceFilteredCached(
+	restype APIResource,
+	ns string,
+	name string,
+	label labels.Selector,
+) ([]*corev1.ObjectReference, error) {
+	ls, err := api.getLister(restype)
+	if err != nil {
+		return nil, err
+	}
+
+	objs, err := ls.ByNamespace(ns).List(label)
+	if err != nil {
+		return nil, err
+	}
+
+	refs := []*corev1.ObjectReference{}
+	for _, obj := range objs {
+		objMeta, ok := obj.(*metav1.PartialObjectMetadata)
+		if !ok {
+			return nil, fmt.Errorf("couldn't convert obj %v to PartialObjectMetadata", obj)
+		}
+		gvk, err := restype.GVK()
+		if err != nil {
+			return nil, err
+		}
+		ref := &corev1.ObjectReference{
+			Kind:            gvk.Kind,
+			APIVersion:      gvk.GroupVersion().String(),
+			Name:            objMeta.GetName(),
+			Namespace:       objMeta.GetNamespace(),
+			UID:             objMeta.GetUID(),
+			ResourceVersion: objMeta.GetResourceVersion(),
+		}
+		refs = append(refs, ref)
+	}
+
+	return refs, nil
+}
+
+// GetOwnerKindAndName returns the pod owner's kind and name, using owner
+// references from the Kubernetes API. The kind is represented as the
+// Kubernetes singular resource type (e.g. deployment, daemonset, job, etc.).
+// When the shared informer cache doesn't return anything we try again with a
+// direct Kubernetes API call.
+func (api *MetadataAPI) GetOwnerKindAndName(ctx context.Context, pod *corev1.Pod) (string, string, error) {
+	ownerRefs := pod.GetOwnerReferences()
+	if len(ownerRefs) == 0 {
+		// pod without a parent
+		return "pod", pod.Name, nil
+	} else if len(ownerRefs) > 1 {
+		log.Debugf("unexpected owner reference count (%d): %+v", len(ownerRefs), ownerRefs)
+		return "pod", pod.Name, nil
+	}
+
+	parent := ownerRefs[0]
+	var parentObj metav1.Object
+	var err error
+	switch parent.Kind {
+	case "Job":
+		parentObj, err = api.getByNamespaceCached(Job, pod.Namespace, parent.Name)
+		if err != nil {
+			log.Warnf("failed to retrieve job from indexer %s/%s: %s", pod.Namespace, parent.Name, err)
+			parentObj, err = api.Client.BatchV1().Jobs(pod.Namespace).Get(ctx, parent.Name, metav1.GetOptions{})
+			if err != nil {
+				log.Warnf("failed to retrieve job from direct API call %s/%s: %s", pod.Namespace, parent.Name, err)
+			}
+		}
+	case "ReplicaSet":
+		var rsObj metav1.Object
+		rsObj, err = api.getByNamespaceCached(RS, pod.Namespace, parent.Name)
+		isNil := rsObj.(*metav1.PartialObjectMetadata) == nil
```

Comment:
I think we can simplify this if `rsObj` is declared as a `PartialObjectMetadata` in the first place.